VSDiagnostics/VSDiagnostics/VSDiagnostics/Diagnostics/Async/SyncMethodWithAsyncSuffix/SyncMethodWithAsyncSuffixAnalyzer.cs

```diff
@@ -0,0 +1,109 @@
+﻿using System.Collections.Immutable;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using VSDiagnostics.Utilities;
+
+namespace VSDiagnostics.Diagnostics.Async.SyncMethodWithSyncSuffix
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class SyncMethodWithAsyncSuffixAnalyzer : DiagnosticAnalyzer
+    {
+        private const DiagnosticSeverity Severity = DiagnosticSeverity.Warning;
+
+        private static readonly string Category = VSDiagnosticsResources.AsyncCategory;
+        private static readonly string Message = VSDiagnosticsResources.SyncMethodWithSyncSuffixAnalyzerMessage;
+        private static readonly string Title = VSDiagnosticsResources.SyncMethodWithSyncSuffixAnalyzerTitle;
+
+        internal static DiagnosticDescriptor Rule
+            =>
+                new DiagnosticDescriptor(DiagnosticId.SyncMethodWithSyncSuffix, Title, Message, Category, Severity,
+                    isEnabledByDefault: true);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.RegisterSyntaxNodeAction(AnalyzeSyntaxNode, SyntaxKind.MethodDeclaration);
+        }
+
+        private static void AnalyzeSyntaxNode(SyntaxNodeAnalysisContext context)
+        {
+            var method = context.Node as MethodDeclarationSyntax;
+            if (method == null)
+            {
+                return;
+            }
+
+            if (method.Modifiers.Any(SyntaxKind.OverrideKeyword))
+            {
+                return;
+            }
+
+            var returnType = context.SemanticModel.GetTypeInfo(method.ReturnType);
+            if (returnType.Type == null)
+            {
+                return;
+            }
+
+            var declaredSymbol = context.SemanticModel.GetDeclaredSymbol(method);
+            if (declaredSymbol == null)
+            {
+                return;
+            }
+
+            if (IsDefinedInAncestor(declaredSymbol))
+            {
+                return;
+            }
+
+            if (!(method.Modifiers.Any(SyntaxKind.AsyncKeyword) ||
+                returnType.Type.MetadataName == typeof(Task).Name ||
+                returnType.Type.MetadataName == typeof(Task<>).Name))
+            {
+                if (method.Identifier.Text.EndsWith("Async"))
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(Rule, method.Identifier.GetLocation(),
+                        method.Identifier.Text));
+                }
+            }
+        }
+
+        private static bool IsDefinedInAncestor(IMethodSymbol methodSymbol)
```

Comment:
This is the same method as in https://github.com/Vannevelj/VSDiagnostics/blob/ccf2ba76e6cb93873b297bee6970dd184ef7a5ff/VSDiagnostics/VSDiagnostics/VSDiagnostics/Diagnostics/Async/AsyncMethodWithoutAsyncSuffix/AsyncMethodWithoutAsyncSuffixAnalyzer.cs#L75 right? These should be moved to an extension method in util/Extensions.cs so we don't duplicate this behaviour. This is even more important because if I recall correctly, there is a scenario in which this method fails (but I haven't looked too deep into it yet -- there's a bug report filed).